### PR TITLE
Implementation of scalable DataBuffer

### PR DESCRIPTION
**Pull request by lballes**
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/renate/updaters/experimental/gdumb.py [hunk 1]
```diff
@@ -56,9 +56,10 @@ def __init__(
 
     def load_state_dict(self, model: RenateModule, state_dict: Dict[str, Any], **kwargs) -> None:
         """Restores the state of the learner."""
+        if not hasattr(self, "_memory_buffer"):
+            self._memory_buffer = GreedyClassBalancingBuffer()
```
Comment (wistuba) [inline src/renate/updaters/experimental/gdumb.py:60]: is this happening because we create an instance without actually calling the constructor?

File: test/renate/updaters/experimental/test_er.py [hunk 2]
```diff
@@ -81,11 +81,10 @@ def test_er_validation_buffer(tmpdir):
     )
     for i in range(3):
         for j in range(100):
-            assert torch.all(
-                torch.eq(
-                    datasets_val[i][j][0],
-                    model_updater._learner._val_memory_buffer[i * 100 + j][0][0],
-                )
+            assert torch.allclose(
```
Comment (wistuba) [inline test/renate/updaters/experimental/test_er.py:84]: why is `allclose` needed? tolerance is pretty high

File: src/renate/memory/buffer.py [hunk 3]
```diff
@@ -197,189 +136,204 @@ def set_transforms(
...
-        storage_mode: defaults.SUPPORTED_BUFFER_STORAGE_MODES = defaults.BUFFER_STORAGE_MODE,
-        transform: Optional[Callable] = None,
-        target_transform: Optional[Callable] = None,
-    ) -> None:
-        super().__init__(
-            storage_mode=storage_mode, transform=transform, target_transform=target_transform
-        )
-        self._max_size = 16
+    def save(self, target_dir: str) -> None:
+        if len(self):
```
Comment (wistuba) [inline src/renate/memory/buffer.py:165]: meaning of `len(self)` is not clear

File: src/renate/memory/buffer.py [hunk 4]
```diff
@@ -161,34 +84,50 @@ def __getitem__(self, idx: int) -> Tuple[NestedTensors, DataDict]:
...
 
-    @abstractmethod
-    def _update(self, dataset: Dataset, metadata: DataDict) -> None:
-        pass
+    def set_metadata(self, key: str, values: torch.Tensor) -> None:
+        if values.size(0) != len(self):
+            raise ValueError()
+        if key not in self._metadata:
+            self._add_metadata_like({key: values})
+        self._metadata[key][: len(self)] = values.cpu()
```
Comment (wistuba) [inline src/renate/memory/buffer.py:130]: class docstring claims that devices are not changed

File: src/renate/memory/buffer.py [hunk 5]
```diff
@@ -197,189 +136,204 @@ def set_transforms(
...
         self._max_size = state_dict["max_size"]
-        self._storage_mode = state_dict["storage_mode"]
         self._seed = state_dict["seed"]
         self._count = state_dict["count"]
-        self._size = state_dict["size"]
-        self._data_points = state_dict["data_points"]
-        self.metadata = state_dict["metadata"]
-
+        self._indices = state_dict["indices"]
+        self._data_point_prototype = state_dict["data_point_prototype"]
```
Comment (wistuba) [inline src/renate/memory/buffer.py:160]: what if storing data is not allowed?

File: src/renate/memory/buffer.py [hunk 6]
```diff
@@ -197,189 +136,204 @@ def set_transforms(
...
-                    ).item()
-                ]
-                self[memory_idx_to_replace] = (
-                    dataset[i],
-                    {key: value[i] for key, value in metadata.items()},
-                )
-                self._class_to_index_map[largest_class].remove(memory_idx_to_replace)
-                self._record_class_index(dataset[i], memory_idx_to_replace)
-
+                largest_class, max_length = self._get_largest_class()
```
Comment (wistuba) [inline src/renate/memory/buffer.py:308]: not very efficient (compute and code). one could keep a counter around (array mapping class index to count) and then use argmax.

File: src/renate/updaters/experimental/gdumb.py (hunk 1 above)
Comment (lballes) [reply src/renate/updaters/experimental/gdumb.py:60]: Yes, as you can see in the old version, we already had to create a new buffer object in `load_state_dict` for precisely that reason. The `if` clause makes sure this only happens when we re-create the learner, not when we're just reloading an old state. (That's necessary because the new buffer implementation relies on persistence of its internal state when `load_state_dict` is called without `load`.)

File: src/renate/memory/buffer.py [hunk 7]
```diff
@@ -104,53 +33,47 @@ class DataBuffer(Dataset, ABC):
...
-
-        self._data_points = None
-        self.metadata = None
-        self._size = 0
-
-        self._transform = transform
-        self._target_transform = target_transform
+        self._datasets = []
+        self._indices = {}
+        self._data_point_prototype = None
```
[the remark this answers is not in the document]
Comment (lballes) [reply src/renate/memory/buffer.py:65]: We  keep around a prototypical data point to infer shape and data type. That is because the memory-mapped tensors that we use for storage don't save that information. The prototype just becomes part of the buffer's state dict, so we can re-infer this info.

File: src/renate/memory/buffer.py (hunk 4 above)
Comment (lballes) [reply src/renate/memory/buffer.py:130]: I guess that passage of the doc string is obsolete. Technically the buffer is now agnostic to the device placement of the data (not the metadata). Metadata is always kept on CPU, so it makes sense to ensure this here so that people can input metadata values (e.g., logits) that reside on GPU.

File: src/renate/memory/buffer.py (hunk 6 above)
Comment (lballes) [reply src/renate/memory/buffer.py:308]: Agreed, but let's fix separately, okay?

**Review by wistuba — APPROVED** [review]
